get_mvn.go

```diff
@@ -0,0 +1,161 @@
+package getter
+
+import (
+	"encoding/xml"
+	"fmt"
+	"io/ioutil"
+	"log"
+	"net/url"
+	"os"
+	"path/filepath"
+	"strings"
+)
+
+// MvnGetter is a Getter implementation that will download an artifact from maven repository, e.g. Sonatype Nexus,
+// uri format: mvn::http://[username@]hostname[:port]/directoryname[?options]
+type MvnGetter struct {
+	HttpGet HttpGetter
+}
+
+func (g *MvnGetter) ClientMode(u *url.URL) (ClientMode, error) {
+	return ClientModeFile, nil
+}
+
+func (g *MvnGetter) Get(dst string, u *url.URL) error {
+	q := u.Query()
+	artifactId := q.Get("artifactId")
+	if artifactId == "" {
+		return fmt.Errorf("query parameter 'artifactId' is required.")
+	}
+	version := q.Get("version")
+	if version == "" {
+		return fmt.Errorf("query parameter 'version' is required.")
+	}
+	classifier := q.Get("classifier")
+
+	artType := q.Get("type")
+	if artType == "" {
+		artType = "jar"
+	}
+
+	filename := artifactId + "-" + version
+	if classifier != "" {
+		filename += "-" + classifier
+	}
+	filename += "." + artType
+
+	dstFile := filepath.Join(dst, filename)
+	return g.GetFile(dstFile, u)
+}
+
+// Get the remote file.
+// The created filename will always be '<artifactId>-<version>[-<classifier>].<type>', Ex., 'testng-6.13.1.jar'. So the name in the `client` passed arg 'dst' is ignored.
+//
+// If the version is a snapshot version, it will get the latest snapshot artifact.
+// Query parameters:
+//   - groupId: the group id
+//   - artifactId: the artifact id
+//   - version: the artifact version
+//   - type: the artifact type, default as 'jar'
+// example url: mvn::http://username@host/mavan/repo/path?groupId=org.example&artifactId=test&version=1.0.0-SNAPSHOT
+func (g *MvnGetter) GetFile(dst string, u *url.URL) error {
+	q := u.Query()
+	groupId := q.Get("groupId")
+	if groupId == "" {
+		return fmt.Errorf("query parameter 'groupId' is required.")
+	}
+	artifactId := q.Get("artifactId")
+	if artifactId == "" {
+		return fmt.Errorf("query parameter 'artifactId' is required.")
+	}
+	version := q.Get("version")
+	if version == "" {
+		return fmt.Errorf("query parameter 'version' is required.")
+	}
+	classifier := q.Get("classifier")
+	artType := q.Get("type")
+	if artType == "" {
+		artType = "jar"
+	}
+
+	artifactUrl, err := url.Parse(u.String())
+	if err != nil {
+		return err
+	}
+	artifactUrl.RawQuery = ""
+	artifactUrl.Path += fmt.Sprintf("/%s/%s/%s", strings.Replace(groupId, ".", "/", -1), artifactId, version)
+
+	ver := version
+	if strings.HasSuffix(version, "-SNAPSHOT") {
+		// get the latest snapshot
+		snapshotVer, err := g.ParseLastestSnapshotVersion(artifactUrl)
+		if err != nil {
+			return err
+		}
+
+		ver = snapshotVer
+	}
+
+	filename := artifactId + "-" + ver
+	if classifier != "" {
+		filename += "-" + classifier
+	}
+	filename += "." + artType
+	artifactUrl.Path += "/" + filename
+	dstFile := dst
```

Comment:
Same here. I don't see a reason for the secondary `dstFile` variable.

Reply:
Also is there a way to use a small helper func to format the filename since there's similar logic here and in Get?